src/libstd/alloc.rs

```diff
@@ -13,15 +13,77 @@
 #![unstable(issue = "32838", feature = "allocator_api")]
 
 #[doc(inline)] #[allow(deprecated)] pub use alloc_crate::alloc::Heap;
-#[doc(inline)] pub use alloc_crate::alloc::{Global, oom};
+#[doc(inline)] pub use alloc_crate::alloc::{Global, Layout, oom};
 #[doc(inline)] pub use alloc_system::System;
 #[doc(inline)] pub use core::alloc::*;
 
+use panicking::HOOK_LOCK;
+
+#[derive(Copy, Clone)]
+enum Hook {
+    Default,
+    Custom(*mut (Fn(Layout) -> ! + 'static + Sync + Send)),
+}
+
+static mut HOOK: Hook = Hook::Default;
+
+/// Registers a custom OOM hook, replacing any that was previously registered.
+///
+/// The OOM hook is invoked when an infallible memory allocation fails.
+/// The default hook prints a message to standard error and aborts the
+/// execution, but this behavior can be customized with the [`set_hook`]
+/// and [`take_hook`] functions.
```

Comment:
The hook currently isn't allowed to unwind IIRC. Probably worth documenting.

Reply:
Isn’t it really? Why?

We have accepted an RFC for `oom=panic` for the use case of handling OOM with `catch_unwind`: https://github.com/rust-lang/rust/issues/48043, https://github.com/rust-lang/rfcs/blob/master/text/2116-alloc-me-maybe.md#oompanic

Reply:
It's definitely intended to work, but there's a comment over here claiming we die in MIR transforms if you try to actually do it: https://github.com/rust-lang/rust/blob/master/src/liballoc/alloc.rs#L104-L121

Reply:
I don't actually know why Box still needs a special lang item to allocate though - maybe we can kill off exchange_alloc entirely at this point?

Reply:
Maybe `box foo` syntax needs a lang item? It’s how `Box::new` is defined.

Reply:
> The hook currently isn't allowed to unwind IIRC. Probably worth documenting.

If the hook unwinding crashes the compiler, but the intent is for it to be allowed per the `oom=panic` RFC, it probably makes sense to leave any comment about unwinding out of the documentation.